feat:support tailwindcss v4

SisyphusZheng opened the pull request:
update /www and tailwind-plugin to support tailwind v4.

Thanks to the merge of #3034, the work of integrating and advancing #2903 is promoted.

related #3034 
related #2903 

close #3056
close #2819

--patch-1 
based on review,

1. del compiler and make it more clean and simple,
2.  fix init tailwind part,
3. fix test tailwind part

--patch-2

1. fix the docs 
2. fix the docs of plugin and version

--patch-3
1. Addressed various maintenance tasks (chores)
2. Updated the Tailwind CSS plugin to align with this release. Generated a fresh project using the updated init configuration and 
3. deployed it via Deno Deploy EA. Deployment is functioning correctly.

--patch-4
Move postcss imports to Tailwind section in init script

--patch-5
Refactored Tailwind plugin type definitions by copying PluginOptions properties from @tailwindcss/postcss and adding Fresh-specific exclude option, while exporting types from package entrypoint for better accessibility.

Test Project: https://fresh.fresh-press.deno.net/
Test Project Repository: https://github.com/SisyphusZheng/fresh-tailwindv4
Test Plugin: https://jsr.io/@freshpress/fresh-plugin-tailwind@1.0.0

SisyphusZheng committed:
feat:support tailwindcss v4

SisyphusZheng committed:
update oxide

SisyphusZheng committed:
fix

iuioiua commented on plugin-tailwindcss/src/mod.ts:
```diff
@@ -1,19 +1,19 @@
   options: TailwindPluginOptions = {},
 ): void {
-  let processor: ReturnType<typeof initTailwind> | null;
+  let processor: ReturnType<typeof initTailwind> | null = null;
```
```suggestion
  const instance = await postcss(twPostcss({
    optimize: app.config.mode === "production",
    ...options,
  }));
```
Might this work? If so, we can remove lines 16-17 and `plugin-tailwindcss/src/compiler.ts`. It'd require `tailwind()`  to become async and be `await`ed where used, including documentation.

iuioiua replied:
I also realise we weren't passing `options` from `tailwind()` to `initTailwind()`. This snippet fixes that.

SisyphusZheng replied:
thx so much

iuioiua commented on deno.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "nodeModulesDir": "auto",
```
Should we also add this to `deno.json` in the init script?

SisyphusZheng replied:
yep thx

SisyphusZheng committed:
update mod.ts

Co-authored-by: Asher Gomez <ashersaupingomez@gmail.com>

SisyphusZheng committed:
del compiler and fix init

SisyphusZheng committed:
fix init and test

iuioiua reviewed (changes requested):
Instances of `tailwind()` not being `await`ed still remain in the codebase. E.g. https://github.com/SisyphusZheng/fresh/blob/aac6207306cfcda2405e2761958136fb637cf470/docs/canary/examples/migration-guide.md?plain=1#L62. I suggest searching for "`tailwind(`".

SisyphusZheng committed:
fix md and version

iuioiua commented on www/deno.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tasks": {
-    "start": "deno run -A --watch=static/,routes/,../src,../docs dev.ts",
+    "start": "deno run -A --allow-scripts --watch=static/,routes/,../src,../docs dev.ts",
```
Is `--allow-scripts` required now? Genuine question. I haven't checked. If so, we should consider whether other tasks such as `build` need it too, including those created from the init script.

SisyphusZheng replied:
The oxide script requires --allow-script during deno install, and I can't figure out a way to avoid this warning for the time being.

iuioiua reviewed (approved):
LGTM!

SisyphusZheng committed:
fix

SisyphusZheng committed:
Update deno.lock

SisyphusZheng committed:
fix

SisyphusZheng committed:
fix logo fmt

SisyphusZheng committed:
fix imports for init

SisyphusZheng committed:
Update types.ts

iuioiua commented on plugin-tailwindcss/src/types.ts:
```diff
@@ -1,48 +1,6 @@
-}
+import type { OnTransformOptions } from "fresh/dev";
 
 export interface TailwindPluginOptions {
```
```suggestion
export interface TailwindPluginOptions extends PluginOptions {
```
From postcss

SisyphusZheng replied:
Thx for rechek,  btw, i already check the https://postcss.org/api/, PostCSS only has `PluginOptions` as a parameter type for `PluginCreator`, not as an exportable interface, and this is Fresh-specific, IMO, may not suitable for this

iuioiua replied:
I think a few things need to be addressed:
1. `TailwindPluginOptions` should be exported from `plugin-tailwindcss/src/mod.ts` (package entrypoint). Otherwise, that interface is inaccessible. IMO, the interface should be moved and `plugin-tailwindcss/src/types.ts` removed, but that might conflict with how Marvin wants things structured.
2. `PluginOptions` is defined in `@tailwindcss/postcss`, not `postcss`. But that's okay, we can just copy the `PluginOptions` interface properties, `base` (not sure if it's applicable for us) and `optimize`, to `TailwindPluginOptions`.
3. These interface properties should be documented. E.g. `optimize` defaults to `true` if Fresh is in production mode and `false` otherwise.

SisyphusZheng committed:
Move postcss imports to Tailwind section in init script

Relocates the addition of 'postcss' and '@tailwindcss/postcss' imports to the block where Tailwind-related imports are set, ensuring they are only included when Tailwind is used.

SisyphusZheng committed:
refactor:based on tailwindcss/postcss/type and reexport

SisyphusZheng committed:
fmt

SisyphusZheng committed:
Merge branch 'main' into patch-test

SisyphusZheng committed:
Update deno.lock

SisyphusZheng committed:
Merge branch 'patch-test' of https://github.com/SisyphusZheng/fresh into patch-test

SisyphusZheng committed:
Merge branch 'main' into patch-test

SisyphusZheng committed:
Merge branch 'main' into patch-test

SisyphusZheng committed:
Update deno.lock

SisyphusZheng committed:
Update deno.lock

marvinhagemeister commented on plugin-tailwindcss/src/mod.ts:
```diff
@@ -1,20 +1,26 @@
+export type { TailwindPluginOptions } from "./types.ts";
 
-export function tailwind<T>(
+export async function tailwind<T>(
```
Let's keep the sync variant of before so that we don't have to change this to be `async`.

SisyphusZheng replied:
yep

SisyphusZheng committed:
Merge branch 'main' into patch-test

SisyphusZheng committed:
Make tailwind plugin synchronous and update usage

Refactored the tailwind plugin to be synchronous instead of async, updating its function signature and all usage examples and documentation accordingly. This simplifies integration and removes unnecessary awaits from code and docs.

SisyphusZheng committed:
Merge branch 'patch-test' of https://github.com/SisyphusZheng/fresh into patch-test

SisyphusZheng committed:
Merge branch 'main' into patch-test

SisyphusZheng committed:
chore

SisyphusZheng committed:
Update deno.lock

iuioiua commented on plugin-tailwindcss/src/mod.ts:
```diff
@@ -1,20 +1,26 @@
     async (args) => {
-      if (!processor) processor = initTailwind(app.config, options);
-      const instance = await processor;
       const res = await instance.process(args.text, {
```
Might this now throw since `instance` could be an unresolved promise?

iuioiua replied:
Ignore... I thought `instance` was a promise.

marvinhagemeister committed:
Merge branch 'main' into patch-test

marvinhagemeister committed:
revert unrelated docs upgrade

marvinhagemeister committed:
revert incorrect changes

marvinhagemeister committed:
upgrade tailwind classes

marvinhagemeister committed:
fix: formatting

marvinhagemeister committed:
fix: tw4 styles

marvinhagemeister committed:
chore: remove tw config file

marvinhagemeister reviewed (approved):
LGTM, Thanks for updating the plugin! I've went ahead and updated the code in `www` for v4 too.

marvinhagemeister committed:
fmt
